testing/tools/integration/cluster.py

```diff
@@ -0,0 +1,823 @@
+        """
+        logging.debug("stop_worker(worker={})".format(worker))
+        if isinstance(worker, Runner):
+            # ref to worker
+            r = self.workers.remove(worker)
+        else: # index of worker in self.workers
+            r = self.workers.pop(worker)
+        r.stop()
+        self.dead_workers.append(r)
+        return r
+
+    def restart_worker(self, worker=-1):
+        """
+        Restart a worker via the `respawn` method of a runner, then add the
+        new Runner instance to `workers`.
+        If `worker` is an int, perform this on the Runner instance at `worker`
+        position in the `dead_workers` list.
+        If `worker` is a Runner instance, perform this on that instance.
+        """
+        logging.debug("restart_worker(worker={})".format(worker))
+        if isinstance(worker, Runner):
+            # ref to dead worker instance
+            old_r = worker
+        else: # index of worker in self.dead_workers
+            old_r = self.dead_workers[worker]
+        r = old_r.respawn()
+        self.workers.append(r)
+        self.runners.append(r)
+        r.start()
+        return r
+
+    def stop_workers(self):
+        logging.debug("stop_workers()")
+        for r in self.runners:
+            r.stop()
+        # move all live workers to dead_workers
+        self.dead_workers.extend(self.workers)
+        self.workers = []
+
+    def get_crashed_workers(self):
+        logging.debug("get_crashed_workers()")
+        return filter(lambda r: r.poll() not in (None, 0,-9,-15), self.runners)
+
+    #########
+    # Sinks #
+    #########
+    def stop_sinks(self):
+        logging.debug("stop_sinks()")
+        for s in self.sinks:
+            s.stop()
+
+    def sink_await(self, values, timeout=30, sink=-1):
+        logging.debug("sink_await(values={}, timeout={}, sink={})".format(
+            values, timeout, sink))
+        if isinstance(sink, Sink):
+            pass
+        else:
+            sink = self.sinks[sink]
+        t = SinkAwaitValue(sink, values, timeout)
+        t.start()
+        t.join()
+        if t.error:
+            raise t.error
+        return sink
+
+    def sink_expect(self, expected, timeout=30, sink=-1):
+        logging.debug("sink_expect(expected={}, timeout={}, sink={})".format(
+            expected, timeout, sink))
+        if isinstance(sink, Sink):
+            pass
+        else:
+            sink = self.sinks[sink]
+        t = SinkExpect(sink, expected, timeout)
+        t.start()
+        t.join()
+        if t.error:
+            raise t.error
+        return sink
+
+    ###########
+    # Senders #
+    ###########
+    def add_sender(self, sender, start=False):
+        logging.debug("add_sender(sender={}, start={})".format(sender, start))
+        self.senders.append(sender)
+        if start:
+            sender.start()
+
+    def join_sender(self, sender=-1, timeout=30):
```

Comment:
Not coming from the threaded-programming world, the verb `join` seems alien to me in this context. Especially since the `Cluster` models the domain of a working Wallaroo cluster, it could be misinterpreted to mean that a new sender was *joining* the cluster.
Would something like `wait_for_sender(self, sender,timeout)` be acceptible?

Reply:
That's a good point.